### PR TITLE
feat(core): add dynamic elements (Fe, Material, Semantic)

### PR DESCRIPTION
**Pull request by frontegg-david**

### CODE REVIEW

**Review by frontegg-david — COMMENTED** [review]
reviewed all pull request that has been merged to this branch
@aviadmizrachi  Can you please approve this PR

File: packages/elements-material-ui/src/Table/Table.tsx [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+  }
+  if (props.hasOwnProperty('filters') && !props.onFilterChange) {
+    throw Error('FeTable: you must provide onFilterChange property if filters is controlled');
+  }
+  if (props.hasOwnProperty('pagination') && !props.pageSize) {
+    throw Error('FeTable: you must provide pageSize property if pagination enabled');
+  }
+  if (props.hasOwnProperty('onPageChange') && !props.pageCount) {
+    throw Error('FeTable: you must provide pageCount property if onPageChange is controlled');
+  }
```
Comment (erano067) [inline packages/elements-material-ui/src/Table/Table.tsx:201]: consider export it to validate function

File: packages/elements-material-ui/src/Table/Table.tsx [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+          page={tableState.pageIndex}
+          onChangePage={(e, page) => onPageChangeHandler(page)}
+          ActionsComponent={(props) => (
+            <TablePaginationActions {...props} gotoPage={gotoPage} pageOptions={pageOptions} />
+          )}
+        />
+      )}
+    </Paper>
+  );
+};
```
Comment (erano067) [inline packages/elements-material-ui/src/Table/Table.tsx:330]: should be refactored, too big for one file

**Review by erano067 — COMMENTED** [review]
Please take a look

File: packages/core/src/elements/Popup/interfaces.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { ReactNode, HTMLAttributes, ReactElement } from 'react';
+
+export type PopupPosition = 'tr' | 'tl' | 'tc' | 'br' | 'bl' | 'bc' | 'lc' | 'rc';
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Popup/interfaces.ts:3]: I have discussed it with Max, but this is way we share our props with Material and Semantic

File: packages/core/src/elements/Table/FeTable.tsx [hunk 4]
```diff
@@ -37,42 +42,57 @@ import { FeTableTHead } from './FeTableTHead';
...
+          const row = cell.row as Row<T> & UseRowSelectRowProps<T>;
+          return (
+            <FeCheckbox
+              {...row.getToggleRowSelectedProps()}
+              checked={row.isSelected}
+              onChange={(e) => onRowSelected(row.original, e.target.checked)}
+            />
+          );
+        },
+      });
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Table/FeTable.tsx:90]: You are right.. I will split it

File: packages/core/src/elements/Table/FeTable.tsx [hunk 5]
```diff
@@ -82,6 +102,9 @@ export const FeTable: FC<TableProps> = <T extends object>(props: TableProps<T>)
   if (props.pagination) {
     tableHooks.push(usePagination);
   }
+  if (props.selection) {
+    tableHooks.push(useRowSelect);
+  }
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Table/FeTable.tsx:104]: its not standartd hooks.. it's like plugins to the ReactTable library

File: packages/core/src/elements/Table/FeTableTHead.tsx [hunk 6]
```diff
@@ -4,20 +4,42 @@ import classNames from 'classnames';
...
+    selectedFlatRows,
+    isAllRowsSelected,
+  } = props;
   return (
     <div className='fe-table__thead'>
       {headerGroups.map((headerGroup) => (
         <div className='fe-table__thead-tr' {...headerGroup.getHeaderGroupProps()}>
           {headerGroup.headers.map((c) => {
             const column = c as FeTableColumnProps<T>;
+            if (column.id === 'fe-selection') {
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Table/FeTableTHead.tsx:32]: this is custom component inside the table columns that injected if  using table with multi selection  row

File: packages/core/src/elements/Table/FeTableTHead.tsx [hunk 7]
```diff
@@ -4,20 +4,42 @@ import classNames from 'classnames';
...
+              return (
+                <div className={classNames('fe-table__thead-tr-th')} {...column.getHeaderProps()}>
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Table/FeTableTHead.tsx:34]: okay, i will fix

File: packages/core/src/elements/Checkbox/FeCheckbox.scss [hunk 8]
```diff
@@ -0,0 +1,96 @@
+.fe-checkbox {
+  position: relative;
+  outline: none;
+  display: inline-block;
+
+  &__in-form {
+    margin-bottom: var(--element-spacing);
+    margin-right: calc(var(--element-spacing) * 2);
+  }
+  &__full-width {
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Checkbox/FeCheckbox.scss:10]: it's the same, but in forms it's more powerful to use display block, for preventing any side components

File: packages/core/src/elements/Popup/PopupComponent.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import React, { forwardRef, ReactNode } from 'react';
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Popup/PopupComponent.tsx:1]: removed

File: packages/core/src/elements/Popup/FePopup.tsx [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    return position;
+  }
+  if (top || bottom) {
+    return `${position}${left ? '-start' : right ? '-end' : ''}`;
+  }
+  return position;
+};
+
+export const FePopup = forwardRef<Popup, PopupProps>((props, ref) => {
+  const { position, trigger, action, content } = props;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Popup/FePopup.tsx:38]: added to the popup container props

File: packages/elements-material-ui/src/Table/TableExpandable.tsx [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+import { Box, Collapse, TableCell, TableRow } from '@material-ui/core';
+
+type TableExpandableProps<T extends object> = {
+  isExpanded: boolean;
+  renderExpandedComponent?: (data: T, index: number) => React.ReactNode;
+  row: Row<T>;
+};
+export const TableExpandable: FC<TableExpandableProps<any>> = <T extends object>(props: TableExpandableProps<T>) => {
+  const { isExpanded, renderExpandedComponent, row } = props;
+  const ref = useRef<HTMLDivElement | null>();
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Table/TableExpandable.tsx:12]: removed

File: packages/elements-material-ui/src/Table/TableFilterColumn.tsx [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  column: FeTableColumnProps<T>;
+  onFilterChange?: (column: FeTableColumnProps<T>, value: any) => void;
+};
+
+export const TableFilterColumn: FC<FeTableFilterColumnProps> = <T extends object>({
+  column,
+  onFilterChange,
+}: FeTableFilterColumnProps<T>) => {
+  if (!column.canFilter) {
+    return null;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Table/TableFilterColumn.tsx:27]: moved

File: packages/elements-material-ui/src/Table/Table.tsx (hunk 2 above)
Comment (frontegg-david) [reply packages/elements-material-ui/src/Table/Table.tsx:329]: I will export some of the functionalities to helpers and utils

File: packages/elements-material-ui/src/Table/Table.tsx (hunk 1 above)
Comment (frontegg-david) [reply packages/elements-material-ui/src/Table/Table.tsx:200]: I will export some of the functionalities to helpers and utils

File: packages/core/package.json [hunk 13]
```diff
@@ -24,8 +24,10 @@
     "react": ">16.8.6",
     "react-dom": ">16.8.6",
     "react-i18next": "^11.7.0",
+    "react-popper-tooltip": "^3.1.0",
     "react-redux": "^7.1.1",
     "react-router-dom": "^5.1.2",
+    "react-select": "^3.1.0",
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/package.json:30]: it's okay, why not? this is basic library that should do the best select functionalities for select components

File: packages/core/src/elements/Table/FeTableFilterColumn.tsx [hunk 14]
```diff
@@ -15,23 +17,27 @@ export const FeTableFilterColumn: FC<FeTableFilterColumnProps> = <T extends obje
   if (!column.canFilter) {
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Table/FeTableFilterColumn.tsx:17]: moved

File: packages/core/src/hooks.ts [hunk 15]
```diff
@@ -1,7 +1,22 @@
 import { useTranslation, UseTranslationResponse } from 'react-i18next';
 import { useSelector, useDispatch } from 'react-redux';
 import * as FFormik from 'formik';
+import { useEffect, useState } from 'react';
 
 export { useSelector, useDispatch };
 export { FFormik };
 export const useT = (): UseTranslationResponse => useTranslation();
+
+export default function useDebounce(value: any, delay: number) {
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/hooks.ts:10]: fixed

File: packages/elements-material-ui/package.json [hunk 16]
```diff
@@ -20,9 +20,11 @@
   "dependencies": {
     "@material-ui/core": "^4.11.0",
     "@material-ui/icons": "^4.9.1",
+    "@material-ui/lab": "^4.0.0-alpha.56",
     "classnames": "^2.2.6",
     "react": ">16.8.6",
     "react-is": "^16.13.1",
+    "react-popper-tooltip": "^3.1.0",
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/package.json:27]: what is the problem here?
@erano067

File: packages/elements-material-ui/src/Checkbox/style.scss [hunk 17]
```diff
@@ -0,0 +1,3 @@
+.fe-semantic-checkbox {
+  margin-right: 1rem;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Checkbox/style.scss:2]: removed

File: packages/elements-material-ui/src/Input/index.tsx [hunk 18]
```diff
@@ -5,29 +5,64 @@ import {
...
+  } else if (suffixIcon) {
+    data.InputProps = {
+      endAdornment: (
+        <InputAdornment onClick={iconAction} position='end'>
+          {iconAction ? <IconButton>{suffixIcon}</IconButton> : suffixIcon}
+        </InputAdornment>
+      ),
+    };
+  }
+  return data;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Input/index.tsx:65]: here we have more work after merge.. the Input not finalized

File: packages/elements-material-ui/src/Popup/PopupClick.tsx [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+        anchorOrigin={{
+          vertical: anchorOrigin.vertical,
+          horizontal: anchorOrigin.horizontal,
+        }}
+        transformOrigin={{
+          vertical: transformOrigin.vertical,
+          horizontal: transformOrigin.horizontal,
+        }}
+      >
+        <Box className={classnames(classes.box, 'fe-m-popup-content')}>{content}</Box>
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Popup/PopupClick.tsx:44]: changed

File: packages/elements-material-ui/src/Popup/style.scss [hunk 20]
```diff
@@ -0,0 +1,8 @@
+.material-popup-container {
+  z-index: 10;
+  padding: 16px;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Popup/style.scss:3]: removed

File: packages/elements-material-ui/src/Popup/style.scss [hunk 21]
```diff
@@ -0,0 +1,8 @@
+.material-popup-container {
+  z-index: 10;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Popup/style.scss:2]: removed

File: packages/elements-material-ui/src/Popup/styles.ts [hunk 22]
```diff
@@ -0,0 +1,9 @@
+import { makeStyles, createStyles, Theme } from '@material-ui/core';
+
+export const useStyles = makeStyles((theme: Theme) =>
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Popup/styles.ts:3]: fixed

File: packages/elements-material-ui/src/Select/Select.tsx [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+      loading={loading ?? false}
+      disableCloseOnSelect
+      open={propOpen ?? open}
+      noOptionsText={noOptionsText ?? 'Empty'}
+      loadingText={loadingText ?? 'Loading...'}
+      onOpen={(e) => (onOpen ? onOpen(e) : setOpen(true))}
+      onClose={(e, reson) => (onClose ? onClose(e, reson) : setOpen(false))}
+      onChange={(e, newValue, reson) => handleChange(e, newValue, reson)}
+      getOptionLabel={(option: any) => (getOptionLabel ? getOptionLabel(option) : option.label)}
+      renderTags={(tagValue, getTagProps) => tagValue.map((option, index) => renderTag(option, getTagProps, index))}
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Select/Select.tsx:66]: why tag and not the right Chip here. this is the same library

File: packages/elements-semantic/src/Input/index.tsx [hunk 24]
```diff
@@ -33,8 +35,31 @@ export class Input extends React.Component<InputProps> {
         </label>
       );
     }
-    const inputProps = { ...mapper(rest), label: inputLabel };
 
-    return <InputComponent {...inputProps}>{children}</InputComponent>;
+    if (rest.prefixIcon) {
+      iconContent = rest.prefixIcon;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-semantic/src/Input/index.tsx:40]: should be refactroed after merge

File: packages/elements-material-ui/src/Table/helpers.ts [hunk 25]
```diff
@@ -0,0 +1,15 @@
+import { useEffect, useState } from 'react';
+
+export function useDebounce(value: any, delay: number) {
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Table/helpers.ts:3]: right. will be moved to to core

File: packages/core/src/elements/Popup/FePopup.tsx [hunk 26]
```diff
@@ -0,0 +1,74 @@
+import { PopupPosition, PopupProps } from './interfaces';
+import React, { forwardRef, useMemo, cloneElement, MouseEvent } from 'react';
+import Popup from 'react-popper-tooltip';
+import './FePopup.scss';
+
+const positions: { [key: string]: string } = {
+  t: 'top',
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Popup/FePopup.tsx:7]: refactored

File: packages/core/src/elements/Popup/FePopup.tsx [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+  e: 'end',
+};
+
+const preparePosition = (p?: PopupPosition): any => {
+  if (!p) {
+    return 'bottom';
+  }
+
+  const center = p.indexOf('c') !== 1;
+  const left = p.indexOf('l') !== 1;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Popup/FePopup.tsx:22]: refactored

File: packages/core/src/elements/Popup/interfaces.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Popup/interfaces.ts:3]: fixed

File: packages/core/src/elements/Select/interfaces.tsx [hunk 28]
```diff
@@ -0,0 +1,47 @@
+import { ReactNode } from 'react';
+import { Size, Theme } from '../../styles';
+
+export interface SelectOptionProps<T> {
+  label: string;
+  value: T;
+  // key: string | number; // REMOVE: use value as key for semantic
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Select/interfaces.tsx:7]: fixed

File: packages/core/src/elements/Select/interfaces.tsx [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+
+  noOptionsText?: string; // used for empty rows
+  loadingText?: string; // used for display loading row in menu
+
+  // input props
+  size?: Size;
+  theme?: Theme;
+}
+
+// const props = {
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/core/src/elements/Select/interfaces.tsx:40]: removed

File: packages/elements-material-ui/package.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/package.json:27]: no problem to. use this libs

File: packages/elements-material-ui/src/Popup/types.ts [hunk 30]
```diff
@@ -0,0 +1,11 @@
+import { PopupProps } from '@frontegg/react-core';
+
+type positions = {
+  vertical: any;
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Popup/types.ts:4]: refactored

File: packages/elements-material-ui/src/Select/index.tsx [hunk 31]
```diff
@@ -0,0 +1 @@
+export * from './Select';
```
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Select/index.tsx:1]: moved

File: packages/elements-material-ui/src/Table/helpers.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (frontegg-david) [reply packages/elements-material-ui/src/Table/helpers.ts:3]: fixed

**Review by erano067 — APPROVED** [review]

**Pull request merged**